### PR TITLE
v0.9.0.2

### PR DESCRIPTION
**Pull request by jjr840430**

### CODE REVIEW

File: src/hcl.c [hunk 1]
```diff
@@ -139,26 +168,14 @@ static int parse_options(const ucl_object_t *obj, struct mcfg_hcl_options *opt)
 	}
 
 	/* get switch */
-	sub_obj = ucl_object_lookup(obj, "switch");
+	sub_obj = ucl_object_lookup(obj, "cmdline");
 	if (!sub_obj || sub_obj->type != UCL_OBJECT) {
-		mcfg_set_err("Could not found 'switch' key for config '%s'", opt->cfg_name);
+		mcfg_set_err("Could not found 'cmdline' key for config '%s'", opt->cfg_name);
```
Comment (erikbor) [inline src/hcl.c]: s/found/find

File: src/hcl.c [hunk 2]
```diff
@@ -139,26 +168,14 @@ static int parse_options(const ucl_object_t *obj, struct mcfg_hcl_options *opt)
...
-	sub_obj = ucl_object_lookup(obj, "description");
-	if (!sub_obj || sub_obj->type != UCL_OBJECT) {
-		mcfg_set_err("Could not found 'description' key for config '%s'", opt->cfg_name);
-		return -1;
-	}
-
-	if (parse_desc_options(sub_obj, opt) != 0) {
-		mcfg_set_err("Invalid description options for config '%s'", opt->cfg_name);
+	if (parse_cmdline_options(sub_obj, opt) != 0) {
+		mcfg_set_err("Invalid cmdline options for config '%s'", opt->cfg_name);
```
Comment (erikbor) [inline src/hcl.c]: s/options/option

**Comment by jjr840430** [comment]
@ronaldtse @erikbor I changed `HCL` configuration structure. Can you please check it and let me know your opinions?

The diff is [here](https://github.com/riboseinc/libmulticonfig/pull/3/commits/c9df6569ca4a6d0ccd9a31a99d9fa679263a3592#diff-bb480d5608ad59c31cf291b59e570f23)

File: tests/cfg_test.c [hunk 3]
```diff
@@ -35,12 +35,47 @@
...
+	struct hcl_option *childs;
+	struct hcl_option *next;
+	struct hcl_option *parent;
+
+	void *data;
+}
+
+static struct hcl_option *g_opts;
+
 static void usage(const char *prog_name)
```
Comment (erikbor) [inline tests/cfg_test.c]: can you please remove `prog_name`, and use either `extern const char *__progname;` or `getprogname()` (the latter you used previously)

**Comment by ronaldtse** [comment]
@jjr840430 I checked the updated configuration syntax -- for the HCL configuration we want every block to have a name "ID". The ID allows cross referencing. Could we provide that?

**Comment by jjr840430** [comment]
@ronaldtse Is it not possible to use `key` instead of ID?
In example config, `log_directory`, `log_level` is the key to identify each block.

**Comment by ronaldtse** [comment]
Ah, I think we want `[block-type] [block-id] { ...`?

**Comment by jjr840430** [comment]
What should `block-type` be specified in the config?

**Comment by ronaldtse** [comment]
Maybe right now there's only one block type, called `config_option`? We will certainly have others later?

**Comment by jjr840430** [comment]
@ronaldtse I added `block type`. The diff is [here](https://github.com/riboseinc/libmulticonfig/pull/3/commits/d96e3748aa19503ec4dc1832a87013b06e53a601)

**Comment by ronaldtse** [comment]
How about the `helper` block? We also need to specify its name? We want to link all of them into a `program` block too. i.e. you can define a `program` with its multiple `config_options`?

**Comment by jjr840430** [comment]
@ronaldtse I'm now implementing to link the configurations of `program` to `config_options` block.
I will implement it today and update in this pull request.

**Comment by jjr840430** [comment]
@ronaldtse `libmulticonfig` should process two HCL configurations:

- HCL config with a command line, environment variable, the configuration key([example](https://github.com/riboseinc/libmulticonfig/blob/v0.9.0.2/tests/cmdline.hcl))
- The configuration of `program`([example](https://github.com/riboseinc/libmulticonfig/blob/v0.9.0.2/tests/cfg.hcl))

Do you have suggestions how to name two HCL configurations separately?

**Comment by jjr840430** [comment]
@ronaldtse so, Do I need to keep two configurations in a single configuration file?

**Comment by ronaldtse** [comment]
No, the program metaconfig belongs in the program source code (before packaging); the program config belongs in the users system (after installing package). They should be separate.

**Pull request merged**